clients/src/main/java/org/apache/kafka/clients/consumer/StubbedAsyncKafkaConsumer.java

```diff
@@ -0,0 +1,548 @@
+            ByteBuffer keyBytes = record.key();
+            byte[] keyByteArray = keyBytes == null ? null : Utils.toArray(keyBytes);
+            K key = keyBytes == null ? null : this.keyDeserializer.deserialize(partition.topic(), headers, keyByteArray);
+            ByteBuffer valueBytes = record.value();
+            byte[] valueByteArray = valueBytes == null ? null : Utils.toArray(valueBytes);
+            V value = valueBytes == null ? null : this.valueDeserializer.deserialize(partition.topic(), headers, valueByteArray);
+            return new ConsumerRecord<>(partition.topic(), partition.partition(), offset,
+                    timestamp, timestampType,
+                    keyByteArray == null ? ConsumerRecord.NULL_SIZE : keyByteArray.length,
+                    valueByteArray == null ? ConsumerRecord.NULL_SIZE : valueByteArray.length,
+                    key, value, headers, leaderEpoch);
+        } catch (RuntimeException e) {
+            throw new RecordDeserializationException(partition, record.offset(),
+                    "Error deserializing key/value for partition " + partition +
+                            " at offset " + record.offset() + ". If needed, please seek past the record to continue consumption.", e);
+        }
+    }
+
+    @Override
+    public void commitSync() {
+        commitSync(Duration.ofMillis(defaultApiTimeoutMs));
+    }
+
+    @Override
+    public void commitSync(Duration timeout) {
+        commitSync(subscriptions.allConsumed(), timeout);
+    }
+
+    @Override
+    public void commitSync(Map<TopicPartition, OffsetAndMetadata> offsets) {
+        commitSync(offsets, Duration.ofMillis(defaultApiTimeoutMs));
+    }
+
+    @Override
+    public void commitSync(Map<TopicPartition, OffsetAndMetadata> offsets, Duration timeout) {
+        // The background thread will do the following when it receives this event:
+        //
+        // 1. Metadata will update its 'last seen' epoch, if newer
+        // 2. Invoke the coordinator commitOffsetsSync logic
+        eventHandler.addAndGet(new CommitSyncEvent(offsets), time.timer(timeout));
+    }
+
+    @Override
+    public void commitAsync() {
+        commitAsync(null);
+    }
+
+    @Override
+    public void commitAsync(OffsetCommitCallback callback) {
+        commitAsync(subscriptions.allConsumed(), callback);
+    }
+
+    @Override
+    public void commitAsync(Map<TopicPartition, OffsetAndMetadata> offsets, OffsetCommitCallback callback) {
+        // The background thread will do the following when it receives this event:
+        //
+        // 1. Metadata will update its 'last seen' epoch, if newer
+        // 2. Invoke the coordinator commitOffsetsAsync logic
+        eventHandler.add(new CommitAsyncEvent(offsets, callback));
+    }
+
+    @Override
+    public void seek(TopicPartition partition, long offset) {
+        // TODO: This needs to be propagated to the background thread, right?
```

Comment:
Yes I think so, and also the next `poll` should respect this call, hence the `poll` call itself should block until the background thread has processed this event and updated its fetch position accordingly, also the returned records from the raw records buffer should be trimmed based on the new position as well.